File: src/learning_service_cpp/src/sum_server.cpp
/**
 * 需求：编写服务端 解析客户端请求的两个数据 将其相加 响应给客户端
 * 流程：
 *      1、包含头文件
 *      2、初始化 ROS2 客户端
 *      3、自定义节点类
 *          3-1、创建服务端
 *          3-2、解析并客户端的请求 并输出响应结果
 *      4、调用 spin 函数 并传入节点对象指针
 *      5、释放资源
 */

#include "base_interfaces/srv/add_ints.hpp"
#include "rclcpp/rclcpp.hpp"

using base_interfaces::srv::AddInts;

using std::placeholders::_1;
using std::placeholders::_2;

class SumServer : public rclcpp::Node {
public:
    SumServer() : Node("sum_server_cpp") {
        RCLCPP_INFO(this->get_logger(), "服务端创建成功，等待客户端请求...");
        server_ = this->create_service<AddInts>("sum", std::bind(&SumServer::sum_callback, this, _1, _2));
    }

private:
    void sum_callback(const AddInts::Request::SharedPtr req, const AddInts::Response::SharedPtr res) {
        res->sum = req->num1 + req->num2;
        RCLCPP_INFO(
            this->get_logger(),
            "客户端请求的数据: [num1 = %d, num2 = %d], 响应结果为: sum = %d",
            req->num1, req->num2, res->sum
        );
    }

    rclcpp::Service<AddInts>::SharedPtr server_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SumServer>());
    rclcpp::shutdown();

    return 0;
}

File: src/learning_service_py/learning_service_py/sum_server.py
"""
需求：编写服务端 提取客户端请求的两个整型数据 并将其相加作为响应
步骤：
    1、导包
    2、初始化 ROS2 客户端
    3、自定义节点类
        3-1、创建服务端
        3-2、编写回调函数 处理请求并做出响应
    4、调用 spin 函数 并传入节点对象
    5、释放资源
"""

import rclpy
from rclpy.node import Node
from base_interfaces.srv import AddInts

class SumServer(Node):
    def __init__(self):
        super().__init__("sum_server_py")
        self.get_logger().info("服务端创建成功!")
        self.server_ = self.create_service(AddInts, "sum", self.sum_callback)

    def sum_callback(self, request, response):
        response.sum = request.num1 + request.num2
        self.get_logger().info(
            f"客户端发送的请求为: [num1 = {request.num1}, num2 = {request.num2}]\n"
            f"服务端作出的响应为: [sum = {response.sum}]"
        )
        return response

def main():
    rclpy.init()
    rclpy.spin(SumServer())
    rclpy.shutdown()

if __name__ == "__main__":
    main()